lib/hexpm/emails/bamboo.ex

```diff
@@ -14,14 +18,38 @@ defmodule Hexpm.Emails.Bamboo.SESAdapter do
 
     message = ExAws.SES.build_message(email.html_body, email.text_body, email.subject)
 
-    ExAws.SES.send_email(destination, message, email(email.from), [])
-    |> ExAws.request!()
+    request = ExAws.SES.send_email(destination, message, email(email.from), [])
+    send_email(request, email, 0)
   end
 
   def handle_config(config) do
     config
   end
 
+  defp send_email(request, email, times)  do
+    request
+    |> ExAws.request()
+    |> maybe_retry(request, email, times)
+  end
+
+  defp maybe_retry({:error, {:http_error, 454, _body}} = error, request, email, times) do
```

Comment:
We need this because hexpm_billing will share the rate limit with hexpm and I don't want to couple them by sending emails through one of the apps.

Reply:
👍 

note, we're throttling in another place too: https://github.com/hexpm/hexpm/blob/master/lib/hexpm/emails/mailer.ex#L11. Have you thought about unifying it in one place?

There's a couple of options:
1. convert `{:error, {:http_error, 454, ...` into some `{:error, :throttling_failure}` that Mailer understands and will retry.
2. Move `Hexpm.Throttle.wait` and chunking to the adapter
3. Do nothing

either option is fine by me :D

Reply:
I think they are different. One is proactively throttling based on known limits, the other is reactively retrying if we are being throttled.

They retrying needs to happen on this level because this is the only place we have access to the HTTP response and lifting an adapter specific error to the application seems wrong. Maybe we should move the wait throttling here, it's hard to say if it's an application concern or an SES concern. I think I will leave it where it is now but you make a good point, we might move it in the future.

Reply:
In the future I will extract the common functionality of hexpm and hexpm_billing, the email logic is duplicated right now. When we do that we can look at refactoring the throttling.